File: src/sendmail.py
#!/usr/bin/python3
import argparse
import configparser
import json
import logging
import os
import sys
import urllib.error
from email.parser import Parser
from email.utils import parseaddr
from html import escape
from mimetypes import guess_type
from urllib import request

TG_MAX_TEXT_LENGTH = 4096

config = configparser.ConfigParser()
config_filepaths = [
    'sendmail.ini', '/etc/tg-sendmail.ini', '/etc/sendmail.ini',
]
for config_filepath in config_filepaths:
    if os.path.exists(config_filepath):
        config.read(config_filepath)
        break
else:
    raise ValueError('Cannot find configuration file')
main_config = config['main']
telegram_config = config['telegram']

logging.basicConfig(
    level=main_config['log_level'].upper(),
    filename=main_config['log_path'],
    filemode='a',
    format='{asctime} {name} - {levelname} - {message}',
    style='{',
)
logger = logging.getLogger('sendmail')


def get_updates(token):  # noqa: C901
    url = f'https://api.telegram.org/bot{token}/getUpdates'
    try:
        response = request.urlopen(url).read().decode()
    except urllib.error.URLError as error:
        logger.error('get updates error. URL: %s Error: %s', url, error)
        return
    logger.debug('get updates response: %s', response)
    try:
        updates = json.loads(response)
    except json.JSONDecodeError as error:
        logger.error('get updates json decode error: %s', error)
        return
    if not updates['ok']:
        logger.error('get updates response is not ok')
        return
    messages = updates['result']
    if not messages:
        warning_message = 'No updates. Try to send some messages to the bot'
        logger.warning(warning_message)
        sys.stdout.write(warning_message)
        return
    for message in messages:
        logger.debug('get updates message: %s', json.dumps(message))
        chat_id = message['message']['chat']['id']
        username = message['message']['chat']['username']
        log_string = f'Chat Id: {chat_id}. Username: {username}'
        logger.info('get updates data: %s', log_string)
        sys.stdout.write(log_string)


def generate_full_name(full_name, address):
    full_address = address
    if full_name:
        full_address = f'{full_name} <{address}>'
    return full_address


def parse_args():
    parser = argparse.ArgumentParser(
        description='Drop-in replacement for any MTA sendmail',
    )
    parser.add_argument(
        '--send-file',
        help='Send local file to the telegram bot',
    )
    parser.add_argument(
        '--get-updates',
        action='store_true',
        help='Run getUpdates method with bot token to achieve chat_id',
    )
    parser.add_argument('-F', help='Set the full name of the sender.')
    parser.add_argument(
        '-f',
        help='Sets the name of the ''from'' person (i.e., the envelope sender'
             ' of the mail). This address may also be used in the From: header'
             ' if that header is missing during initial submission.',
    )
    parser.add_argument(
        '-t', action='store_true', help='Read message for recipients.',
    )
    parser.add_argument('remains', nargs='*', help='Should be an email')
    return parser.parse_known_args()


def prepare_email(sender_name, sender_address, remains, parse_to_header):
    email = Parser().parse(sys.stdin)
    if not parse_to_header and not email['to'] and remains:
        for arg in remains:
            full_name, address = parseaddr(arg)
            # https://datatracker.ietf.org/doc/html/rfc822#section-6
            # due to this rfc we actually don't care about validness of an
            # address, just put any of given information into `To:` if it is
            # empty
            email.add_header('To', generate_full_name(full_name, address))
    if not email['from'] and sender_address:
        email.add_header(
            'From',
            generate_full_name(sender_name, sender_address),
        )
    return email


def generate_message(email):
    message = [f'Node: <b>{os.uname().nodename}</b>\n']
    for header, value in email.items():
        message.append(f'<i>{escape(header)}:</i> <b>{escape(value)}</b>')
    message.append('\n')
    message.append(f'<pre>{escape(email.get_payload())}</pre>')
    return '\n'.join(message)


def send(message, token, chat_id):
    long_message = ''
    url = f'https://api.telegram.org/bot{token}/sendMessage'
    if len(message) > TG_MAX_TEXT_LENGTH:
        long_message = message
        message = (
            f'{message[:500]}\n\n'
            '<b>Message is too long. See attached file below</b>'
        )
    payload = {
        'chat_id': chat_id,
        'text': message,
        'parse_mode': 'HTML',
    }
    logger.debug('Send payload: %s', json.dumps(payload))
    headers = {
        'Content-Type': 'application/json',
    }
    data = json.dumps(payload).encode()
    req = request.Request(url, data, headers)
    with request.urlopen(req) as response:
        res = response.read()
        logger.debug('Response: %s', res.decode())
    if long_message:
        send_file("long_message.txt", long_message.encode(), token, chat_id)


def send_file(filepath, content, token, chat_id):
    body = encode_multipart_formdata(filepath, content)
    print(body)
    headers = {
        'Content-Type': 'multipart/form-data; boundary=boundary',
    }
    url = f'https://api.telegram.org/bot{token}/sendDocument?chat_id={chat_id}'
    req = request.Request(url, body, headers)
    with request.urlopen(req) as response:
        res = response.read()
        logger.debug('Response: %s', res.decode())


def read_content(filepath):
    with open(filepath, 'rb') as file:
        return file.read()


def get_content_type(filepath):
    mimetype, _ = guess_type(filepath)
    return mimetype or 'application/octet-stream'


def encode_multipart_formdata(filepath: str, content: bytes):
    return b'\r\n'.join([
        b'--boundary',
        b'Content-Disposition: form-data; name="document"; '
        b'filename="%s"' % os.path.basename(filepath).encode(),
        b'Content-Type: %s' % get_content_type(filepath).encode(),
        b'',
        b'%s' % content,
        b'--boundary--',
    ])


if __name__ == '__main__':
    chat_id = telegram_config['chat_id']
    bot_token = telegram_config['bot_token']
    valid_credentials = chat_id and bot_token
    args, unknown_args = parse_args()
    if args.get_updates:
        # used to get know chat_id
        get_updates(bot_token)
        exit()
    elif args.send_file:
        if not os.path.exists(args.send_file):
            raise ValueError(f'File {args.send_file} does not exists!')
        if not valid_credentials:
            logger.warning('Please fill /etc/tg-sendmail.ini configuration file!')
            exit(1)
        send_file(args.send_file, read_content(args.send_file), bot_token, chat_id)
        exit()
    email = prepare_email(args.F, args.f, args.remains, args.t)
    logger.debug('Prepared email: %s', email.as_string())
    message = generate_message(email)
    logger.debug('Prepared message: %s', message)
    if not valid_credentials:
        logger.warning('Please fill /etc/tg-sendmail.ini configuration file!')
        exit(1)
    send(message, bot_token, chat_id)

File: sendmail.go
package main

import (
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"html"
	"io"
	"log"
	"mime"
	"net/http"
	"net/mail"
	"os"
	"path/filepath"
	"strings"

	"gopkg.in/yaml.v3"
)

const (
	TGMaxTextLength = 4096
)

type Config struct {
	Log struct {
		Level string `yaml:"level"`
		Path  string `yaml:"path"`
	} `yaml:"log"`
	Telegram struct {
		BotToken string `yaml:"bot_token"`
		ChatID   string `yaml:"chat_id"`
	} `yaml:"telegram"`
}

var (
	config Config
	logger *log.Logger
)

func init() {
	// Try to load config from multiple locations
	configPaths := []string{
		"sendmail.yaml",
		"/etc/tg-sendmail.yaml",
		"/etc/sendmail.yaml",
	}

	var configFound bool
	for _, path := range configPaths {
		if _, err := os.Stat(path); err == nil {
			if err := loadConfig(path); err == nil {
				configFound = true
				break
			}
		}
	}

	if !configFound {
		fmt.Fprintln(os.Stderr, "Cannot find configuration file")
		os.Exit(1)
	}

	// Setup logging
	logFile, err := os.OpenFile(config.Log.Path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0666)
	if err != nil {
		log.Fatal(err)
	}
	logger = log.New(logFile, "", log.LstdFlags)
}

func loadConfig(path string) error {
	data, err := os.ReadFile(path)
	if err != nil {
		return err
	}

	return yaml.Unmarshal(data, &config)
}

func getUpdates(token string) {
	url := fmt.Sprintf("https://api.telegram.org/bot%s/getUpdates", token)
	resp, err := http.Get(url)
	if err != nil {
		logger.Printf("get updates error. URL: %s Error: %v\n", url, err)
		return
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		logger.Printf("get updates read error: %v\n", err)
		return
	}

	logger.Printf("get updates response: %s\n", string(body))

	var result struct {
		Ok     bool `json:"ok"`
		Result []struct {
			Message struct {
				Chat struct {
					ID       int64  `json:"id"`
					Username string `json:"username"`
				} `json:"chat"`
			} `json:"message"`
		} `json:"result"`
	}

	if err := json.Unmarshal(body, &result); err != nil {
		logger.Printf("get updates json decode error: %v\n", err)
		return
	}

	if !result.Ok {
		logger.Println("get updates response is not ok")
		return
	}

	if len(result.Result) == 0 {
		warningMessage := "No updates. Try to send some messages to the bot"
		logger.Println(warningMessage)
		fmt.Print(warningMessage)
		return
	}

	for _, update := range result.Result {
		chatID := update.Message.Chat.ID
		username := update.Message.Chat.Username
		logString := fmt.Sprintf("Chat Id: %d. Username: %s", chatID, username)
		logger.Printf("get updates data: %s\n", logString)
		fmt.Print(logString)
	}
}

func generateFullName(fullName, address string) string {
	if fullName != "" {
		return fmt.Sprintf("%s <%s>", fullName, address)
	}
	return address
}

func prepareEmail(senderName, senderAddress string, remains []string, parseToHeader bool) (*mail.Message, error) {
	msg, err := mail.ReadMessage(os.Stdin)
	if err != nil {
		return nil, err
	}

	if !parseToHeader && msg.Header.Get("To") == "" && len(remains) > 0 {
		for _, arg := range remains {
			addr, err := mail.ParseAddress(arg)
			if err != nil {
				// If parsing fails, use as-is
				msg.Header["To"] = append(msg.Header["To"], arg)
			} else {
				msg.Header["To"] = append(msg.Header["To"], generateFullName(addr.Name, addr.Address))
			}
		}
	}

	if msg.Header.Get("From") == "" && senderAddress != "" {
		msg.Header["From"] = []string{generateFullName(senderName, senderAddress)}
	}

	return msg, nil
}

func generateMessage(msg *mail.Message) (string, error) {
	hostname, err := os.Hostname()
	if err != nil {
		hostname = "unknown"
	}

	var message strings.Builder
	message.WriteString(fmt.Sprintf("Node: <b>%s</b>\n\n", html.EscapeString(hostname)))

	for key, values := range msg.Header {
		for _, value := range values {
			message.WriteString(fmt.Sprintf("<i>%s:</i> <b>%s</b>\n",
				html.EscapeString(key), html.EscapeString(value)))
		}
	}

	message.WriteString("\n")

	body, err := io.ReadAll(msg.Body)
	if err != nil {
		return "", err
	}

	message.WriteString(fmt.Sprintf("<pre>%s</pre>", html.EscapeString(string(body))))

	return message.String(), nil
}

func send(message, token, chatID string) error {
	longMessage := ""
	if len(message) > TGMaxTextLength {
		longMessage = message
		message = message[:500] + "\n\n<b>Message is too long. See attached file below</b>"
	}

	payload := map[string]interface{}{
		"chat_id":    chatID,
		"text":       message,
		"parse_mode": "HTML",
	}

	jsonData, err := json.Marshal(payload)
	if err != nil {
		return err
	}

	logger.Printf("Send payload: %s\n", string(jsonData))

	url := fmt.Sprintf("https://api.telegram.org/bot%s/sendMessage", token)
	resp, err := http.Post(url, "application/json", bytes.NewBuffer(jsonData))
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	respBody, _ := io.ReadAll(resp.Body)
	logger.Printf("Response: %s\n", string(respBody))

	if longMessage != "" {
		return sendFile("long_message.txt", []byte(longMessage), token, chatID)
	}

	return nil
}

func sendFile(filepath string, content []byte, token, chatID string) error {
	body, contentType := encodeMultipartFormdata(filepath, content)

	url := fmt.Sprintf("https://api.telegram.org/bot%s/sendDocument?chat_id=%s", token, chatID)
	req, err := http.NewRequest("POST", url, bytes.NewBuffer(body))
	if err != nil {
		return err
	}

	req.Header.Set("Content-Type", contentType)

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	respBody, _ := io.ReadAll(resp.Body)
	logger.Printf("Response: %s\n", string(respBody))

	return nil
}

func readContent(filepath string) ([]byte, error) {
	return os.ReadFile(filepath)
}

func getContentType(filepath string) string {
	mimeType := mime.TypeByExtension(filepath)
	if mimeType == "" {
		return "application/octet-stream"
	}
	return mimeType
}

func encodeMultipartFormdata(filepath string, content []byte) ([]byte, string) {
	boundary := "boundary"
	var buf bytes.Buffer

	buf.WriteString("--" + boundary + "\r\n")
	buf.WriteString(fmt.Sprintf("Content-Disposition: form-data; name=\"document\"; filename=\"%s\"\r\n",
		filepath))
	buf.WriteString(fmt.Sprintf("Content-Type: %s\r\n", getContentType(filepath)))
	buf.WriteString("\r\n")
	buf.Write(content)
	buf.WriteString("\r\n")
	buf.WriteString("--" + boundary + "--")

	return buf.Bytes(), fmt.Sprintf("multipart/form-data; boundary=%s", boundary)
}

func main() {
	// Define flags
	sendFileFlag := flag.String("send-file", "", "Send local file to the telegram bot")
	getUpdatesFlag := flag.Bool("get-updates", false, "Run getUpdates method with bot token to achieve chat_id")
	senderFullName := flag.String("F", "", "Set the full name of the sender.")
	senderAddress := flag.String("f", "", "Sets the name of the 'from' person (i.e., the envelope sender of the mail).")
	parseToHeader := flag.Bool("t", false, "Read message for recipients.")
	
	// Parse known flags and collect remaining args
	flag.Parse()
	remains := flag.Args()

	// Handle -oi flag (ignore single dot on line) - standard sendmail compatibility
	// Just consume it silently for compatibility
	for i, arg := range os.Args[1:] {
		if arg == "-oi" {
			// Remove from remains if present
			newRemains := []string{}
			for j, r := range remains {
				if j != i {
					newRemains = append(newRemains, r)
				}
			}
			remains = newRemains
			break
		}
	}

	chatID := config.Telegram.ChatID
	botToken := config.Telegram.BotToken
	validCredentials := chatID != "" && botToken != ""

	if *getUpdatesFlag {
		getUpdates(botToken)
		return
	}

	if *sendFileFlag != "" {
		if _, err := os.Stat(*sendFileFlag); os.IsNotExist(err) {
			fmt.Fprintf(os.Stderr, "File %s does not exist!\n", *sendFileFlag)
			os.Exit(1)
		}
		if !validCredentials {
			logger.Println("Please fill /etc/tg-sendmail.ini configuration file!")
			os.Exit(1)
		}
		content, err := readContent(*sendFileFlag)
		if err != nil {
			logger.Printf("Error reading file: %v\n", err)
			os.Exit(1)
		}
		if err := sendFile(filepath.Base(*sendFileFlag), content, botToken, chatID); err != nil {
			logger.Printf("Error sending file: %v\n", err)
			os.Exit(1)
		}
		return
	}

	email, err := prepareEmail(*senderFullName, *senderAddress, remains, *parseToHeader)
	if err != nil {
		logger.Printf("Error preparing email: %v\n", err)
		os.Exit(1)
	}

	logger.Printf("Prepared email with headers: %v\n", email.Header)

	message, err := generateMessage(email)
	if err != nil {
		logger.Printf("Error generating message: %v\n", err)
		os.Exit(1)
	}

	logger.Printf("Prepared message: %s\n", message)

	if !validCredentials {
		logger.Println("Please fill /etc/tg-sendmail.ini configuration file!")
		os.Exit(1)
	}

	if err := send(message, botToken, chatID); err != nil {
		logger.Printf("Error sending message: %v\n", err)
		os.Exit(1)
	}
}
